R/parse-datetime.r

```diff
@@ -11,30 +11,32 @@
 
 #' @param date contains the date to be formatted
 #' @param mode gives the desired output format (see above)
-datetime_fn <- function(createdat, mode="since", permissive=FALSE) {
+#' @param permissive logical. If false, will error with an improper type.  If true, will just warn.
+datetime_fn <- function(createdat, mode = "since", permissive = FALSE) {
   
   # track which records are NA before the mungebit
   originally_NA <- is.na(createdat)
   
   # do the conversion
-  datetime <- suppressWarnings(lubridate:::ymd_hms(createdat))
+  datetime <- suppressWarnings(lubridate::ymd_hms(createdat))
   if (any(is.na(datetime))) {
     if (mode == 'hod') { stop('Cannot extract hour from date-only.') }
-    datetime <- suppressWarnings(lubridate:::ymd(createdat))
+    datetime <- suppressWarnings(lubridate::ymd(createdat))
   }
   
   # check to see if any records are converted to NA
-  if (any(is.na(datetime[!originally_NA])) && !isTRUE(permissive)) { 
-    stop('Improper date format.') 
+  if (any(is.na(datetime[!originally_NA]))) {
+    error_method <- if (isTRUE(permissive)) warning else stop
+    error_method('Improper date format.')
   }
 
   suppressWarnings(switch(mode,
     "since" = as.numeric(as.Date(datetime)),
-    "hod" = lubridate:::hour(datetime),
+    "hod" = lubridate::hour(datetime),
     "dow" = weekdays(datetime),
-    "dom" = lubridate:::day(datetime),
+    "dom" = lubridate::day(datetime),
     "doy" = syberiaMungebits:::get.doy(datetime),
-    "moy" = lubridate:::month(datetime),
+    "moy" = lubridate::month(datetime),
     "holiday" = syberiaMungebits:::is.holiday(datetime),
```

Comment:
Actually, it looks like this matters significantly -- `as.Date` is incredibly slow, so you shouldn't need to call it more than once. 

```r
> system.time(as.Date(x))
   user  system elapsed
 10.878   0.041  11.279
```

That's only on a 300k length vector, and we might be doing this multiple times in the dataset.